posthog/api/query.py

```diff
@@ -116,6 +117,75 @@ def create(self, request, *args, **kwargs) -> Response:
             capture_exception(e)
             raise
 
+    @extend_schema(
+        request=QueryRequest,
+        responses={
+            200: QueryResponseAlternative,
+        },
+    )
+    @monitor(feature=Feature.QUERY, endpoint="query_eventsource", method="POST")
+    @action(methods=["POST"], detail=False)
+    def eventsource(self, request, *args, **kwargs) -> StreamingHttpResponse:
+        data = self.get_model(request.data, QueryRequest)
+        if data.filters_override is not None:
+            data.query = apply_dashboard_filters_to_dict(
+                data.query.model_dump(), data.filters_override.model_dump(), self.team
+            )  # type: ignore
+
+        if data.variables_override is not None:
+            query_as_dict = data.query.model_dump()
+
+            data.query = apply_dashboard_variables_to_dict(query_as_dict, data.variables_override, self.team)  # type: ignore
+
+        client_query_id = data.client_query_id or uuid.uuid4().hex
+        execution_mode = execution_mode_from_refresh(data.refresh)
+
+        self._tag_client_query_id(client_query_id)
+
+        if execution_mode == execution_mode.CACHE_ONLY_NEVER_CALCULATE:
+            # Here in query endpoint we always want to calculate if the cache is stale
+            execution_mode = ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE
+
+        # with event source we always want to "block" if we're calculating
+        if execution_mode in [
+            ExecutionMode.RECENT_CACHE_CALCULATE_ASYNC_IF_STALE,
+            ExecutionMode.RECENT_CACHE_CALCULATE_ASYNC_IF_STALE_AND_BLOCKING_ON_MISS,
+            ExecutionMode.EXTENDED_CACHE_CALCULATE_ASYNC_IF_STALE,
+        ]:
+            execution_mode = ExecutionMode.RECENT_CACHE_CALCULATE_BLOCKING_IF_STALE
+        if execution_mode in [ExecutionMode.CALCULATE_ASYNC_ALWAYS]:
+            execution_mode = ExecutionMode.CALCULATE_BLOCKING_ALWAYS
```

Comment:
Hmm, so now actually I'm not sure the event streaming approach is feasible. What it means is just a long-running Django connection – and the fundamental reason we implemented async querying with polling, is that our infra _doesn't allow Django connections longer than 60 s_.

Even though the event streaming approach is nicer for conveying progress, it's not actually any different in being just a long Django connection. So unless something has changed on the infra side, this doesn't work for queries over 60 s.

(`livestream` doesn't have this problem, being a separate Golang service.)